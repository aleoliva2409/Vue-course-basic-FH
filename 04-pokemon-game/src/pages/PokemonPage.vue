<template>
  <h1 v-if="!pokemon">Wait a minute please ...</h1>

  <div v-else>
    <h1>Who's that pokemon?</h1>

    <!-- img -->
    <PokemonPicture :pokemonId="pokemon.id" :showPokemon="showPokemon" />

    <!-- options -->
    <!-- el metodo dentro de pokemon-selected(checkAnswer) recibe el como argumento lo que se emita, ejemplo @pokemon-selected="checkAnswer(evento emitido)" -->
    <PokemonOptions @pokemon-selected="checkAnswer" :pokemons="pokemons" />

    <template v-if="showAnswer">
      <h2 class="fade-in">{{ message }}</h2>
      <button class="btn" @click="newGame">New game</button>
    </template>
  </div>
</template>

<script>
import PokemonPicture from '@/components/PokemonPicture';
import PokemonOptions from '@/components/PokemonOptions';

import getPokemonOptions from '@/helpers/getPokemonOptions';

// console.log(getPokemonOptions())

export default {
  components: {
    PokemonPicture,
    PokemonOptions,
  },
  data() {
    return {
      pokemons: [],
      pokemon: null,
      showPokemon: false,
      showAnswer: false,
      message: '',
    };
  },
  methods: {
    async mixPokemons() {
      this.pokemons = await getPokemonOptions();
      const randomId = Math.floor(Math.random() * 4);
      this.pokemon = this.pokemons[randomId];
    },
    checkAnswer(id) {
      if (id) {
        this.showPokemon = true;
        this.showAnswer = true;

        if (id === this.pokemon.id) {
          this.message = `Right!! It's ${this.pokemon.name}`;
        } else {
          this.message = `Oops, you're wrong!!! It was ${this.pokemon.name}`;
        }
      }
    },
    newGame() {
      (this.pokemons = []),
        (this.pokemon = null),
        (this.showPokemon = false),
        (this.showAnswer = false),
        (this.message = ''),
        this.mixPokemons();
    },
  },
  mounted() {
    this.mixPokemons();
  },
};
</script>

<style lang="css" scoped>
.btn {
  font-size: 18px;
  border-radius: 15px;
  padding: 3px 6px;
  border: 1px solid black;
}
</style>
