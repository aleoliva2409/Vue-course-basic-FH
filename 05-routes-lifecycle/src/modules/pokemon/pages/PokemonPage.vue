<template>
  <h1>Pokemon: #{{ id }}</h1>
  <div v-if="pokemon">
    <img :src="pokemon.sprites.front_default" :alt="pokemon.name" />
  </div>
</template>

<script>
export default {
  props: {
    id: {
      type: Number,
      required: true,
    },
  },
  data() {
    return {
      pokemon: null,
    };
  },
  created() {
    this.getPokemon();
  },
  methods: {
    async getPokemon() {
      try {
        const res = await fetch(`https://pokeapi.co/api/v2/pokemon/${this.id}`);
        const pokemon = await res.json();
        this.pokemon = pokemon;
      } catch (error) {
        this.$router.push('/');
      }
    },
  },
  watch: {
    id() {
      this.getPokemon();
    },
  },
};
</script>
