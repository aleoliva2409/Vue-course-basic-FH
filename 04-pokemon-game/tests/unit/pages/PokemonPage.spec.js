import { shallowMount } from '@vue/test-utils';
import PokemonPage from '@/pages/PokemonPage';
import PokemonOptions from '@/components/PokemonOptions';
import PokemonPicture from '@/components/PokemonPicture';
import { pokemons } from '../mocks/pokemons.mock';

describe('PokemonPage Component', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(PokemonPage);
  });

  test('should match with snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('should call mixPokemons when it was mounted', () => {
    const mixPokemonArraySpy = jest.spyOn(PokemonPage.methods, 'mixPokemons');

    shallowMount(PokemonPage);

    expect(mixPokemonArraySpy).toHaveBeenCalled();
  });

  test('should match with snapshot when mixPokemons were loaded', () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemons,
          pokemon: pokemons[0],
          showPokemon: false,
          showAnswer: false,
          message: '',
        };
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  test('should show PokemonPicture and PokemonOptions components', () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemons,
          pokemon: pokemons[0],
          showPokemon: false,
          showAnswer: false,
          message: '',
        };
      },
    });

    const pokemonPicture = wrapper.findComponent(PokemonPicture);
    //? otra opcion
    // const { pokemonId } = pokemonPicture.props()
    const pokemonOptions = wrapper.findComponent(PokemonOptions);

    expect(pokemonPicture.exists()).toBeTruthy();
    expect(pokemonOptions.exists()).toBeTruthy();
    expect(pokemonPicture.vm.pokemonId).toBe(5);
    expect(pokemonOptions.vm.pokemons).toEqual(pokemons);
  });

  test('tests on checkAnswer', async () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemons,
          pokemon: pokemons[0],
          showPokemon: false,
          showAnswer: false,
          message: '',
        };
      },
    });

    await wrapper.vm.checkAnswer(5);

    expect(wrapper.find('h2').exists()).toBeTruthy();
    expect(wrapper.vm.showPokemon).toBe(true);
    expect(wrapper.find('h2').text()).toBe(`Right!! It's ${pokemons[0].name}`);

    await wrapper.vm.checkAnswer(10);

    expect(wrapper.vm.message).toBe(`Oops, you're wrong!!! It was ${pokemons[0].name}`);
  });
});
