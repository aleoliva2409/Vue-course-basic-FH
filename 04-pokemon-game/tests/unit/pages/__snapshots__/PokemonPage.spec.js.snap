// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage Component should match with snapshot 1`] = `<h1>Wait a minute please ...</h1>`;

exports[`PokemonPage Component should match with snapshot when mixPokemons were loaded 1`] = `
<div>
  <h1>Who's that pokemon?</h1><!-- img -->
  <pokemon-picture-stub pokemonid="5" showpokemon="false"></pokemon-picture-stub><!-- options -->
  <!-- el metodo dentro de pokemon-selected(checkAnswer) recibe el como argumento lo que se emita, ejemplo @pokemon-selected="checkAnswer(evento emitido)" -->
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
  <!--v-if-->
</div>
`;
